hal/src/sercom/uart.rs

```diff
@@ -11,242 +11,234 @@ use hal::blocking::serial::{write::Default, Write};
 use hal::serial;
 use nb;
 
-macro_rules! uart_pinout {
-    ([$($Type:ident:
-        ($pad0:ident, $pad1:ident, $pad2:ident, $pad3:ident),)+
-    ]) => {
-$(
-/// Similar to SPI Pinout the UART allows selecting any pad for RX and either pad 0 or 2 for TX
-pub enum $Type {
-    /// Construct pinout with rx assigned to pad0,
-    /// TX here must be pad 2
-    /// One entry, can't put both on pad 0
-    Rx0Tx2{rx: $pad0, tx: $pad2},
-
-    Rx1Tx0{rx: $pad1, tx: $pad0},
-    Rx1Tx2{rx: $pad1, tx: $pad2},
-
-    /// One entry, can't put both on pad 2
-    Rx2Tx0{rx: $pad2, tx: $pad0},
-
-    Rx3Tx0{rx: $pad3, tx: $pad0},
-    Rx3Tx2{rx: $pad3, tx: $pad2},
+pub trait RxpoTxpo {
+    fn rxpo_txpo(&self) -> (u8, u8);
 }
 
-impl $Type {
-    /// Return the txpo and rxpo values for
-    /// this pinout configuration
-    fn rxpo_txpo(&self) -> (u8, u8) {
-        match self {
-            &$Type::Rx0Tx2{..} => (0, 1),
-
-            &$Type::Rx1Tx0{..} => (1, 0),
-            &$Type::Rx1Tx2{..} => (1, 1),
-
-            &$Type::Rx2Tx0{..} => (2, 0),
+macro_rules! uart {
+    ($Type:ident: ($Sercom:ident, $SERCOM:ident, $powermask:ident, $clock:ident)) => {
+        $crate::paste::item! {
+            /// A padout configuration for the SERCOM in UART mode.
+            pub struct [<$Type Padout>]<RX, TX, RTS, CTS> {
+                pub rx: RX,
+                pub tx: TX,
+                pub rts: RTS,
+                pub cts: CTS,
+            }
+        }
 
-            &$Type::Rx3Tx0{..} => (3, 0),
-            &$Type::Rx3Tx2{..} => (3, 1),
+        macro_rules! padout {
+            ($rxpo_txpo:expr => $pad0:ident, $pad1:ident) => {
+                $crate::paste::item! {
+                    impl<PIN0, PIN1> From<([<$Sercom $pad0>]<PIN0>, [<$Sercom $pad1>]<PIN1>)> for [<$Type Padout>]<[<$Sercom $pad0>]<PIN0>, [<$Sercom $pad1>]<PIN1>, (), ()> {
```

Comment:
these impls could also do with a brief comment to help future grok-ability

Reply:
:+1: